### ISSUE TITLE
Inquiry About Gradient Flow in Learnable Threshold Implementation

### ISSUE THREAD

**Issue by RicePasteM** [Hiyuur/SR-TOD#5]
Thank you for sharing your interesting work and implementation! While studying the code for SRTOD_CascadeRCNN, I noticed a potential optimization issue that I'd like to clarify regarding the learnable threshold mechanism.
The current implementation uses torch.sign when creating the difference map mask:
```
difference_map_mask = (torch.sign(difference_map - learnable_thresh) + 1)*0.5
```
As `torch.sign` has zero gradients everywhere (non-differentiable at zero), this appears to block gradient flow to the `learnable_thresh` parameter, effectively preventing it from being updated during training despite being declared as a `nn.Parameter`.
Could you please clarify:
- Whether this gradient blockage is an intentional design choice?
- If there's an alternative implementation strategy you would recommend?
- Any empirical observations about the threshold's actual learning behavior in your experiments?

Your insights would be invaluable for better understanding the design rationale.

**Comment by Hiyuur** [comment]
Thank you for your very meaningful question. I apologize for replying only now as I was busy preparing for my internship. Initially, I tried using torch.where for binarization, but I received a warning about gradient being blocked during backpropagation. This is due to the non-differentiable nature of torch.where. Therefore, I tried to use activation functions for numerical computation instead of torch.where. When I used the simplest torch.sign, the warning about gradient being blocked disappeared, and in my experimental observations, the learnable threshold parameters indeed undergo slight updates (usually showing a gradual decrease). Perhaps there are some approximate differentiable mechanisms in torch? If you want to use a more explicit strategy, you might consider using scaled versions of torch.tanh or torch.sigmoid to approximate binarization (like difference_map_mask = torch.sigmoid(100 * (difference_map - learnable_thresh)) ). You could also try to implement a custom backpropagation function for this part to achieve approximate differentiability. Thank you again for your question.

**Comment by RicePasteM** [comment]
Thank you for the detailed explanation. Using scaled torch.tanh/sigmoid or custom backpropagation sounds great. I'll give them a try. Appreciate your insights!
